Sources/SwiftIfConfig/IfConfigEvaluation.swift

```diff
@@ -0,0 +1,516 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+import SwiftSyntax
+import SwiftOperators
+
+enum IfConfigError: Error, CustomStringConvertible {
```

Comment:
I definitely agree that we should be able to form `Diagnostic` instances. Given the need for warnings, I'm thinking about providing diagnostics (not errors), and if there will be a way for the APIs to throw, we can throw `DiagnosticsError`. We could do something similar to the approach of `SwiftOperators` and have a "diagnostic handler" that we stream out diagnostics to. It can be allowed to throw, and will make all of the APIs `rethrows` to account for cases where the client decides to throw.

But it would be good to settle on some kind of pattern here.